src/environmentd/src/http.rs

```diff
@@ -271,6 +275,43 @@ impl AuthedClient {
         &mut self,
         request: HttpSqlRequest,
     ) -> Result<SqlHttpExecuteResponse, anyhow::Error> {
+        // This API prohibits executing statements with responses whose
+        // semantics are at odds with an HTTP response.
+        fn prohibit_stmts(stmt: &Statement<Raw>, results: &mut Vec<SimpleResult>) -> bool {
+            let execute_responses = Plan::generated_from(stmt.into())
+                .into_iter()
+                .map(ExecuteResponse::generated_from)
+                .flatten()
+                .collect::<Vec<_>>();
+
+            if execute_responses.iter().any(|execute_response| {
+                matches!(
+                    execute_response,
+                    ExecuteResponseKind::Fetch
+                        | ExecuteResponseKind::SetVariable
+                        | ExecuteResponseKind::Tailing
+                        | ExecuteResponseKind::CopyTo
+                        | ExecuteResponseKind::CopyFrom
+                        | ExecuteResponseKind::Raise
+                        | ExecuteResponseKind::DeclaredCursor
+                        | ExecuteResponseKind::ClosedCursor
+                )
+            }) && !matches!(
+                stmt,
+                // Both `SelectStatement` and `CopyStatement` generate
```

Comment:
Can ExecuteResponseKind be taught about this instead so that we don't have to poke at the statement type? I'd prefer to use one style of approach instead of both. (Honestly, I'm not opposed to exclusively sniffing the statement type and using that instead of the ExecuteResponseKind stuff.)

Reply:
If we only sniff `Statement`, any statements growing the ability to generate invalid `ExecuteResponse`s will still have to either execute and generate an error (or panic, I guess).

I'm not sure through what means you could teach `ExecuteResponse` about the ability to be acceptable iff it's generated from a certain type of `Statement`. Do you have a suggestion?

Reply:
ping

Reply:
This is a good concern. Can we do both? Specifically: during parsing into statement(s), do an initial check for unsupported responses and error (preventing any statements from running). Then if we encounter one later, we also error (preventing the current and next transactions from committing).